### PR TITLE
Added gradient boosting binding

### PR DESCRIPTION
**Pull request by TirelessClock**
I noticed that mlpack doesn't currently have a binding for Gradient Boosting. 

To give a brief description of Gradient Boosting: it's an ensemble technique which uses weak learners such as shallow Decision Trees (or Decision stumps). Each learner is trained to predict the errors, or gradients, of previous learners, and their subsequent result is combined to give the final result. 

Gradient Boosting is a very common and very powerful ensemble technique used for a wide variety of reasons, from classification to regression. More significantly, Gradient Boosting is a pathway to implementing more powerful algorithms, primarily XGBoost, which is just an optimized version of Gradient Boosting. For these reasons I implemented Gradient Boosting's binding.  

I've implemented Gradient Boosting in `mlpack/src/mlpack/methods/grad_boosting`.

Files added:
- `grad_boosting_main.cpp`: Main binding file. 
- `grad_boosting.hpp`: The base grad_boosting class. Only has the interface for the methods, not their implementation.
- `grad_boosting_impl.hpp`: Implementation of the grad_boosting class. 
- `grad_boosting_model.hpp`: The serializable grad_boosting model class. Only has the interface for the methods, not their implementation.
- `grad_boosting_model_impl.hpp`: Implementation of serializable grad_boosting model.

Please note: this PR is still incomplete and I have a few more changes before it can be finalized, but I've made the PR now so I can include it in my GSoC'24 proposal.

### CODE REVIEW

**Comment by shrit** [comment]
@TirelessClock the code looks good, I skimmed the implementation rapidly, but I will go more into detail.
Sorry for my slow review on this one. Could you add a couple more of tests ?

**Comment by TirelessClock** [comment]
@shrit Yeah sure. Thanks for the review!

**Comment by rcurtin** [comment]
If we are going to add a C++ class for it too, do you think you can write some documentation for it?  Take a look at `doc/user/methods/decision_tree.md` for an example; the structure can be adapted.

**Comment by TirelessClock** [comment]
@rcurtin Yeah, I'll start work on the documentation + testing immediately.

File: src/mlpack/methods/grad_boosting/grad_boosting_model.hpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        const arma::mat& testData,
+        arma::Row<size_t>& predictions
+    );
+
+    //! Classify test points.
+    void Classify(
+        const arma::mat& testData,
+        arma::Row<size_t>& predictions,
+        arma::mat& probabilities
+    );
```
Comment (shrit) [inline src/mlpack/methods/grad_boosting/grad_boosting_model.hpp:103]: Please follow mlpack style in here, regarding the parameters.

File: src/mlpack/methods/grad_boosting/grad_boosting_model.hpp [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        const arma::mat& data,
+        const arma::Row<size_t>& labels,
+        const size_t numClasses,
+        const size_t num_models
+    );
+
+    //! Classify test points.
+    void Classify(
+        const arma::mat& testData,
+        arma::Row<size_t>& predictions
```
Comment (shrit) [inline src/mlpack/methods/grad_boosting/grad_boosting_model.hpp:95]: @TirelessClock any reason why you did not add `MatType` template to this function ?

File: src/mlpack/methods/grad_boosting/grad_boosting_model.hpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        arma::mat& probabilities
+    );
+
+    //! Serialize the model.
+    template<typename Archive>
+    void serialize(Archive& ar, const uint32_t /* version */) {
+        if (cereal::is_loading<Archive>()) {
+            delete dsBoost;
+            dsBoost = NULL;
+        }
```
Comment (shrit) [inline src/mlpack/methods/grad_boosting/grad_boosting_model.hpp:111]: The same here, the `{` goes on a new line after if statement or a function implementation

**Review by shrit — CHANGES_REQUESTED** [review]
@TirelessClock please apply mlpack code style, and add the tests and documentations as discussed.
For more information about the style check the wiki in github, or check other implementation code, which I think it is going to be faster.

**Comment by TirelessClock** [comment]
> @TirelessClock please apply mlpack code style, and add the tests and documentations as discussed. For more information about the style check the wiki in github, or check other implementation code, which I think it is going to be faster.

Hello @shrit , I'm so sorry about the delay, currently a bit overloaded with assignments and exams. I'll finish this up and push by the end of the week.

File: src/mlpack/methods/grad_boosting/grad_boosting_model.hpp (hunk 2 above)
Comment (TirelessClock) [reply src/mlpack/methods/grad_boosting/grad_boosting_model.hpp:95]: @shrit No specific reason. I used MatType in grad_boosting, but used barebone arm::mat in the grad_boosting_model. I guess I made the mistake once and just went with it without realizing.
